File: backend/app/schemas/assessment.py
"""通用评测 (Assessment) 域 Pydantic schemas。

覆盖:名册、评测身份/版本、投放 session、追加式评分、gradebook 与 item 统计。
服务层多数返回 dict 投影(避免异步惰性加载),故这些 schema 直接对 dict 校验。
"""
from datetime import datetime
from decimal import Decimal
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


# --------------------------------------------------------------------------- #
# 学生 / 班级名册
# --------------------------------------------------------------------------- #
class StudentCreateRequest(BaseModel):
    student_no: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    user_id: Optional[int] = None


class StudentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subject_id: int
    user_id: Optional[int] = None
    student_no: str
    name: str


class StudentPage(BaseModel):
    items: List[StudentRead]
    total: int
    page: int
    page_size: int


class ClassroomCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)


class ClassroomRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subject_id: int
    name: str


class ClassroomMembersReplaceRequest(BaseModel):
    student_ids: List[int] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# 评测身份 / 版本
# --------------------------------------------------------------------------- #
class AssessmentCreateRequest(BaseModel):
    """从一个 shared 组稿定稿版本 + 班级创建评测(身份 + v1 + 投放)。"""
    title: str = Field(min_length=1, max_length=255)
    composition_version_id: int
    classroom_id: int
    session_name: Optional[str] = Field(default=None, max_length=255)


class AssessmentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subject_id: int
    title: str
    description: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime


class AssessmentVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    assessment_id: int
    version_no: int
    composition_version_id: Optional[int] = None
    source_type: str
    total_score: Optional[Decimal] = None
    item_count: Optional[int] = None
    created_at: datetime


class AssessmentItemRead(BaseModel):
    """公开条目投影:不含私有 scoring_spec。"""
    id: int
    item_key: str
    position: int
    item_type: str
    max_score: Decimal
    parent_item_id: Optional[int] = None
    source_question_id: Optional[int] = None
    source_question_revision: Optional[int] = None


class AssessmentDetail(AssessmentRead):
    versions: List[AssessmentVersionRead] = Field(default_factory=list)
    current_items: List[AssessmentItemRead] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# 投放 session
# --------------------------------------------------------------------------- #
class AttemptSummary(BaseModel):
    attempt_id: int
    attempt_no: int
    mode: str
    status: str
    revision: int
    total_score: Optional[Decimal] = None
    max_score: Optional[Decimal] = None
    answered_count: int
    graded_count: int


class ParticipationRead(BaseModel):
    id: int
    participant_key: str
    student_id: Optional[int] = None
    user_id: Optional[int] = None
    display_name: Optional[str] = None
    identifier: Optional[str] = None
    attendance_status: str
    status: str
    attempts: List[AttemptSummary] = Field(default_factory=list)


class SessionRead(BaseModel):
    id: int
    assessment_id: int
    assessment_title: Optional[str] = None
    version_id: int
    subject_id: int
    name: str
    delivery_status: str
    grading_status: str
    revision: int
    created_at: datetime
    updated_at: datetime


class SessionDetail(BaseModel):
    id: int
    assessment_id: int
    assessment_title: Optional[str] = None
    version_id: int
    version_no: int
    subject_id: int
    name: str
    delivery_status: str
    grading_status: str
    opens_at: Optional[datetime] = None
    closes_at: Optional[datetime] = None
    attempt_limit: Optional[int] = None
    revision: int
    created_at: datetime
    updated_at: datetime
    total_score: Optional[Decimal] = None
    item_count: Optional[int] = None
    items: List[AssessmentItemRead] = Field(default_factory=list)
    participants: List[ParticipationRead] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# 追加式评分
# --------------------------------------------------------------------------- #
class GradeItemInput(BaseModel):
    item_id: int
    score: Optional[Decimal] = None
    outcome: Optional[str] = None
    grading_method: Optional[str] = None
    feedback: Optional[Any] = None


class GradeAppendRequest(BaseModel):
    """按 attempt 追加评分;expected_revision 乐观锁,batch_id 幂等重试键。"""
    expected_revision: int
    batch_id: Optional[str] = Field(default=None, max_length=64)
    items: List[GradeItemInput] = Field(default_factory=list)

    @field_validator("items")
    @classmethod
    def _non_empty(cls, v: List[GradeItemInput]) -> List[GradeItemInput]:
        if not v:
            raise ValueError("至少需要一条评分项")
        return v


class AttemptGradeRow(BaseModel):
    item_id: int
    item_key: str
    max_score: Decimal
    response_id: Optional[int] = None
    score: Optional[Decimal] = None
    outcome: Optional[str] = None
    grade_revision: int
    status: str


class AttemptGradeResult(BaseModel):
    attempt_id: int
    participation_id: int
    attempt_no: int
    status: str
    revision: int
    total_score: Optional[Decimal] = None
    max_score: Optional[Decimal] = None
    answered_count: int
    graded_count: int
    grades: List[AttemptGradeRow] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# gradebook / 统计
# --------------------------------------------------------------------------- #
class GradebookScore(BaseModel):
    item_id: int
    score: Optional[Decimal] = None
    outcome: Optional[str] = None


class GradebookRow(BaseModel):
    participation_id: int
    participant_key: str
    display_name: Optional[str] = None
    identifier: Optional[str] = None
    attendance_status: str
    participation_status: str
    attempt_id: Optional[int] = None
    attempt_revision: int
    attempt_status: Optional[str] = None
    total_score: Optional[Decimal] = None
    graded_count: int
    scores: List[GradebookScore] = Field(default_factory=list)


class GradebookRead(BaseModel):
    session_id: int
    grading_status: str
    delivery_status: str
    revision: int
    total_score: Optional[Decimal] = None
    item_count: int
    items: List[AssessmentItemRead] = Field(default_factory=list)
    participants: List[GradebookRow] = Field(default_factory=list)
    page: int
    page_size: int
    total: int


class ItemStatistic(BaseModel):
    item_id: int
    item_key: str
    position: int
    max_score: Decimal
    graded_count: int
    average_score: Optional[Decimal] = None
    min_score: Optional[Decimal] = None
    max_score_awarded: Optional[Decimal] = None
    full_marks_count: int
    incorrect_count: int = 0
    # 失分率:未拿满分的比例(相对已评分作答);无评分时为 None。
    error_rate: Optional[Decimal] = None


class ItemStatisticsRead(BaseModel):
    session_id: int
    grading_status: str
    participation_count: int
    items: List[ItemStatistic] = Field(default_factory=list)


# --------------------------------------------------------------------------- #
# 成绩册 Excel 导入
# --------------------------------------------------------------------------- #
class ScoreImportError(BaseModel):
    """单行级别的导入错误(可聚合)。"""
    row: int
    field: str
    message: str


class ScoreImportPreview(BaseModel):
    session_id: int
    valid: bool
    total_rows: int
    changed_rows: int
    changed_scores: int
    unchanged_rows: int
    errors: List[ScoreImportError] = Field(default_factory=list)


class ScoreImportResult(BaseModel):
    session_id: int
    batch_id: Optional[str] = None
    changed_rows: int
    changed_scores: int
    unchanged_rows: int
    applied_attempts: int
    # 幂等重放命中(相同 batch_id + 同文件)时为 True,表示未再次写库。
    replayed: bool = False

File: frontend/app/types/assessment.ts
// 通用评测 (Assessment / Gradebook) 域前端契约 —— 评测重构后的数据层。
//
// 与后端 app/schemas/assessment.py 对齐。注意:后端分值/得分是 Decimal,
// 经 FastAPI(jsonable_encoder)序列化为 JSON **字符串**(如 "40.00")。
// 因此读取投影里的分值一律为 string;写入时后端 Decimal 兼容 number 或 string。

/** 评测身份生命周期。 */
export type AssessmentStatus = 'active' | 'archived'

/** 投放状态。线下默认 closed。 */
export type DeliveryStatus = 'draft' | 'open' | 'closed'

/** 评分状态。finalized 之后只读。 */
export type GradingStatus = 'not_started' | 'in_progress' | 'finalized'

/** 参与者出勤状态。 */
export type AttendanceStatus = 'present' | 'absent' | 'excused'

/** 参与状态。 */
export type ParticipationStatus = 'invited' | 'active' | 'completed' | 'withdrawn'

/** attempt 投放模式。 */
export type AttemptMode = 'offline' | 'online'

/** attempt 状态。 */
export type AttemptStatus =
  | 'not_started'
  | 'in_progress'
  | 'submitted'
  | 'grading'
  | 'graded'
  | 'void'

/** 评分结果分类。 */
export type GradeOutcome = 'correct' | 'partial' | 'incorrect' | 'unscored'

/** 评分方式。 */
export type GradingMethod = 'manual' | 'imported' | 'automatic'

/** 读取投影里的分值(后端 Decimal 序列化为字符串)。 */
export type ScoreValue = string

/** 写入分值:后端 Decimal 兼容 number / string;null 表示清除该题得分。 */
export type ScoreInput = number | string | null

// --------------------------------------------------------------------------- //
// 学生 / 班级名册
// --------------------------------------------------------------------------- //
export interface Student {
  id: number
  subject_id: number
  user_id: number | null
  student_no: string
  name: string
}

export interface Classroom {
  id: number
  subject_id: number
  name: string
}

export interface StudentCreateRequest {
  student_no: string
  name: string
  user_id?: number | null
}

export interface StudentPage {
  items: Student[]
  total: number
  page: number
  page_size: number
}

export interface ClassroomCreateRequest {
  name: string
}

export interface ClassroomMembersReplaceRequest {
  student_ids: number[]
}

// --------------------------------------------------------------------------- //
// 评测身份 / 版本 / 条目
// --------------------------------------------------------------------------- //
export interface Assessment {
  id: number
  subject_id: number
  title: string
  description: string | null
  status: AssessmentStatus | string
  created_at: string
  updated_at: string
}

export interface AssessmentVersion {
  id: number
  assessment_id: number
  version_no: number
  composition_version_id: number | null
  source_type: string
  total_score: ScoreValue | null
  item_count: number | null
  created_at: string
}

/** 公开条目投影:不含私有 scoring_spec。 */
export interface AssessmentItem {
  id: number
  item_key: string
  position: number
  item_type: string
  max_score: ScoreValue
  parent_item_id: number | null
  source_question_id: number | null
  source_question_revision: number | null
}

export interface AssessmentDetail extends Assessment {
  versions: AssessmentVersion[]
  current_items: AssessmentItem[]
}

export interface AssessmentCreateRequest {
  title: string
  composition_version_id: number
  classroom_id: number
  session_name?: string | null
}

// --------------------------------------------------------------------------- //
// 投放 session / 参与 / attempt
// --------------------------------------------------------------------------- //
export interface AttemptSummary {
  attempt_id: number
  attempt_no: number
  mode: AttemptMode | string
  status: AttemptStatus | string
  revision: number
  total_score: ScoreValue | null
  max_score: ScoreValue | null
  answered_count: number
  graded_count: number
}

export interface Participation {
  id: number
  participant_key: string
  student_id: number | null
  user_id: number | null
  display_name: string | null
  identifier: string | null
  attendance_status: AttendanceStatus | string
  status: ParticipationStatus | string
  attempts: AttemptSummary[]
}

export interface Session {
  id: number
  assessment_id: number
  assessment_title: string | null
  version_id: number
  subject_id: number
  name: string
  delivery_status: DeliveryStatus | string
  grading_status: GradingStatus | string
  revision: number
  created_at: string
  updated_at: string
}

export interface SessionDetail {
  id: number
  assessment_id: number
  assessment_title: string | null
  version_id: number
  version_no: number
  subject_id: number
  name: string
  delivery_status: DeliveryStatus | string
  grading_status: GradingStatus | string
  opens_at: string | null
  closes_at: string | null
  attempt_limit: number | null
  revision: number
  created_at: string
  updated_at: string
  total_score: ScoreValue | null
  item_count: number | null
  items: AssessmentItem[]
  participants: Participation[]
}

// --------------------------------------------------------------------------- //
// 追加式评分
// --------------------------------------------------------------------------- //
export interface GradeItemInput {
  item_id: number
  // null 表示清除该题得分。
  score?: ScoreInput
  outcome?: GradeOutcome | string | null
  grading_method?: GradingMethod | string | null
  feedback?: unknown
}

export interface GradeAppendRequest {
  expected_revision: number
  batch_id?: string | null
  items: GradeItemInput[]
}

export interface AttemptGradeRow {
  item_id: number
  item_key: string
  max_score: ScoreValue
  response_id: number | null
  score: ScoreValue | null
  outcome: GradeOutcome | string | null
  grade_revision: number
  status: string
}

export interface AttemptGradeResult {
  attempt_id: number
  participation_id: number
  attempt_no: number
  status: AttemptStatus | string
  revision: number
  total_score: ScoreValue | null
  max_score: ScoreValue | null
  answered_count: number
  graded_count: number
  grades: AttemptGradeRow[]
}

// --------------------------------------------------------------------------- //
// gradebook 读取(分页)
// --------------------------------------------------------------------------- //
export interface GradebookScore {
  item_id: number
  score: ScoreValue | null
  outcome: GradeOutcome | string | null
}

export interface GradebookRow {
  participation_id: number
  participant_key: string
  display_name: string | null
  identifier: string | null
  attendance_status: AttendanceStatus | string
  participation_status: ParticipationStatus | string
  attempt_id: number | null
  attempt_revision: number
  attempt_status: AttemptStatus | string | null
  total_score: ScoreValue | null
  graded_count: number
  scores: GradebookScore[]
}

export interface Gradebook {
  session_id: number
  grading_status: GradingStatus | string
  delivery_status: DeliveryStatus | string
  revision: number
  total_score: ScoreValue | null
  item_count: number
  items: AssessmentItem[]
  participants: GradebookRow[]
  page: number
  page_size: number
  total: number
}

// --------------------------------------------------------------------------- //
// item 统计
// --------------------------------------------------------------------------- //
export interface ItemStatistic {
  item_id: number
  item_key: string
  position: number
  max_score: ScoreValue
  graded_count: number
  average_score: ScoreValue | null
  min_score: ScoreValue | null
  max_score_awarded: ScoreValue | null
  full_marks_count: number
  incorrect_count: number
  // 失分率:未拿满分的比例(相对已评分作答);无评分时为 null。
  error_rate: ScoreValue | null
}

export interface ItemStatistics {
  session_id: number
  grading_status: GradingStatus | string
  participation_count: number
  items: ItemStatistic[]
}

// --------------------------------------------------------------------------- //
// 成绩册 Excel 导入
// --------------------------------------------------------------------------- //
export interface ScoreImportError {
  row: number
  field: string
  message: string
}

export interface ScoreImportPreview {
  session_id: number
  valid: boolean
  total_rows: number
  changed_rows: number
  changed_scores: number
  unchanged_rows: number
  errors: ScoreImportError[]
}

export interface ScoreImportResult {
  session_id: number
  batch_id: string | null
  changed_rows: number
  changed_scores: number
  unchanged_rows: number
  applied_attempts: number
  // 幂等重放命中(相同 batch_id + 同文件)时为 true,表示未再次写库。
  replayed: boolean
}
